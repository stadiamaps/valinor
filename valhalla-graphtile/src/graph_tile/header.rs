use crate::{AsCowStr, BIN_COUNT, GraphId};
use bitfield_struct::bitfield;
use geo::{Coord, coord};
use std::borrow::Cow;
use zerocopy_derive::FromBytes;

/// Remaining variable offset slots for growth.
/// See valhalla/balder/graphtileheader.h for details.
const EMPTY_SLOTS: usize = 11;

#[derive(FromBytes)]
#[bitfield(u64)]
struct FirstBitfield {
    #[bits(46)]
    graph_id: u64,
    #[bits(4)]
    density: u8,
    #[bits(4)]
    name_quality: u8,
    #[bits(4)]
    speed_quality: u8,
    #[bits(4)]
    exit_quality: u8,
    // NOTE: We don't use bool, because that doesn't have a `FromBytes` implementation.
    // We are only using a single bit though, so we *can* guarantee that the bit pattern is valid.
    // This will probably change once we have the ability to assume bit validity.
    // This is currently in beta: https://doc.rust-lang.org/beta/std/mem/trait.TransmuteFrom.html.
    #[bits(1)]
    has_elevation: u8,
    #[bits(1)]
    has_ext_directed_edge: u8,
}

#[derive(FromBytes)]
#[bitfield(u64)]
struct SecondBitfield {
    #[bits(21)]
    node_count: u32,
    #[bits(21)]
    directed_edge_count: u32,
    #[bits(21)]
    predicted_speeds_count: u32,
    #[bits(1)]
    _spare: u8,
}

// NOTE: There are a lot of weird paddings which arose because of an incompatibility in padding
// between x86 and amd64 architectures.
// This was entirely accidental upstream and will eventually be cleaned up in v4.

#[derive(FromBytes)]
#[bitfield(u32)]
struct TransitionCountBitfield {
    #[bits(22)]
    transition_count: u32,
    // Deprecated; scheduled for deletion in v4
    #[bits(10)]
    _spare: u16,
}

#[derive(FromBytes)]
#[bitfield(u32)]
struct TurnLaneCountBitfield {
    #[bits(21)]
    turn_lane_count: u32,
    #[bits(11)]
    _spare: u16,
}

#[derive(FromBytes)]
#[bitfield(u64)]
struct TransitRecordBitfield {
    #[bits(16)]
    transfer_count: u32,
    #[bits(7)]
    _spare: u16,
    #[bits(24)]
    departure_count: u32,
    #[bits(16)]
    stop_count: u16,
    #[bits(1)]
    _spare: u8,
}

#[derive(FromBytes)]
#[bitfield(u64)]
struct MiscCountsBitFieldOne {
    #[bits(12)]
    route_count: u16,
    #[bits(12)]
    schedule_count: u16,
    #[bits(24)]
    sign_count: u32,
    #[bits(16)]
    _spare: u16,
}

#[derive(FromBytes)]
#[bitfield(u64)]
struct MiscCountsBitFieldTwo {
    #[bits(24)]
    access_restriction_count: u32,
    #[bits(16)]
    admin_count: u16,
    #[bits(24)]
    _spare: u32,
}

// TODO: Unsure if packed is needed as well.
// Seems like it's technically not, since Valhalla is explicit about padding
/// Summary information about the graph tile.
///
/// This contains metadata like version,
/// number of nodes and edges,
/// and pointer offsets to other data.
#[derive(Copy, Clone, FromBytes, Debug)]
#[repr(C)]
pub struct GraphTileHeader {
    bit_field_1: FirstBitfield,
    base_lon_lat: [f32; 2],
    version: [u8; 16],
    /// The dataset ID (canonically, the last OSM changeset ID).
    pub dataset_id: u64,
    bit_field_2: SecondBitfield,
    transition_count_bitfield: TransitionCountBitfield,
    turn_lane_count_bitfield: TurnLaneCountBitfield,
    transit_record_bitfield: TransitRecordBitfield,
    misc_counts_bit_field_one: MiscCountsBitFieldOne,
    misc_counts_bit_field_two: MiscCountsBitFieldTwo,
    /// These can be used for custom information.
    /// As long as the size of the header and order of the date of the structure doesn't change,
    /// this should be backward compatible.
    pub reserved: u128,
    // TODO: Valhalla has the unhelpful comment "Offset to the beginning of the variable sized data."
    // We should improve these.
    pub complex_restriction_forward_offset: u32,
    pub complex_restriction_reverse_offset: u32,
    pub edge_info_offset: u32,
    pub text_list_offset: u32,
    /// The date the tile was created (since pivot date).
    /// TODO: Figure out what pivot date means; probably hide the raw value
    pub create_date: u32,
    /// Offsets for each bin of the grid (for search/lookup)
    pub bin_offsets: [u32; BIN_COUNT as usize],
    /// Offset to the beginning of the variable sized data.
    pub lane_connectivity_offset: u32,
    /// Offset to the beginning of the variable sized data.
    pub predicted_speeds_offset: u32,
    /// The size of the tile (in bytes)
    pub tile_size: u32,
    /// See valhalla/balder/graphtileheader.h for details.
    _empty_slots: [u32; EMPTY_SLOTS],
}

impl GraphTileHeader {
    /// The full Graph ID of this tile.
    #[inline]
    pub const fn graph_id(&self) -> GraphId {
        // Safety: We know that the bit field cannot contain a value
        // larger than the max allowed value.
        unsafe { GraphId::from_id_unchecked(self.bit_field_1.graph_id()) }
    }

    /// The relative road density within this tile (0-15).
    #[inline]
    pub const fn density(&self) -> u8 {
        self.bit_field_1.density()
    }

    /// The relative quality of name assignment for this tile (0-15).
    #[inline]
    pub const fn name_quality(&self) -> u8 {
        self.bit_field_1.name_quality()
    }

    /// The relative quality of speed assignment for this tile (0-15).
    #[inline]
    pub const fn speed_quality(&self) -> u8 {
        self.bit_field_1.speed_quality()
    }

    /// The relative quality of exit signs for this tile (0-15).
    #[inline]
    pub const fn exit_quality(&self) -> u8 {
        self.bit_field_1.exit_quality()
    }

    /// Does this tile include elevation data?
    #[inline]
    pub const fn has_elevation(&self) -> bool {
        self.bit_field_1.has_elevation() != 0
    }

    /// Does this tile include extended directed edge attributes?
    #[inline]
    pub const fn has_ext_directed_edge(&self) -> bool {
        self.bit_field_1.has_ext_directed_edge() != 0
    }

    /// The coordinate of the southwest corner of this graph tile.
    #[inline]
    pub const fn sw_corner(&self) -> Coord<f32> {
        coord! {x: self.base_lon_lat[0], y: self.base_lon_lat[1]}
    }

    /// Gets the version of Baldr used to generate this graph tile.
    pub fn version(&self) -> Cow<'_, str> {
        self.version.as_cow_str()
    }

    #[inline]
    pub const fn node_count(&self) -> u32 {
        self.bit_field_2.node_count()
    }

    /// The number of directed edges in this graph tile.
    #[inline]
    pub const fn directed_edge_count(&self) -> u32 {
        self.bit_field_2.directed_edge_count()
    }

    /// The number of predicted speed records in this graph tile.
    #[inline]
    pub const fn predicted_speeds_count(&self) -> u32 {
        self.bit_field_2.predicted_speeds_count()
    }

    /// The number of node transitions in this graph tile.
    #[inline]
    pub const fn transition_count(&self) -> u32 {
        self.transition_count_bitfield.transition_count()
    }

    /// The number of turn lanes in this graph tile.
    #[inline]
    pub const fn turn_lane_count(&self) -> u32 {
        self.turn_lane_count_bitfield.turn_lane_count()
    }

    /// The number of transit transfers in this graph tile.
    #[inline]
    pub const fn transfer_count(&self) -> u32 {
        self.transit_record_bitfield.transfer_count()
    }

    /// The number of transit departures in this graph tile.
    #[inline]
    pub const fn departure_count(&self) -> u32 {
        self.transit_record_bitfield.departure_count()
    }

    /// The number of transit stops in this graph tile.
    #[inline]
    pub const fn stop_count(&self) -> u16 {
        self.transit_record_bitfield.stop_count()
    }

    /// The number of transit routes in this graph tile.
    #[inline]
    pub const fn route_count(&self) -> u16 {
        self.misc_counts_bit_field_one.route_count()
    }

    /// The number of transit schedules in this graph tile.
    #[inline]
    pub const fn schedule_count(&self) -> u16 {
        self.misc_counts_bit_field_one.schedule_count()
    }

    /// The number of signs in this graph tile.
    #[inline]
    pub const fn sign_count(&self) -> u32 {
        self.misc_counts_bit_field_one.sign_count()
    }

    /// The number of access restrictions in this graph tile.
    #[inline]
    pub const fn access_restriction_count(&self) -> u32 {
        self.misc_counts_bit_field_two.access_restriction_count()
    }

    /// The number of admin records in this graph tile.
    #[inline]
    pub const fn admin_count(&self) -> u16 {
        self.misc_counts_bit_field_two.admin_count()
    }
}

#[cfg(test)]
mod test {
    use crate::graph_tile::TEST_GRAPH_TILE;

    #[test]
    fn test_parse_header() {
        let tile = &*TEST_GRAPH_TILE;

        insta::assert_debug_snapshot!(tile.header);
    }
}
