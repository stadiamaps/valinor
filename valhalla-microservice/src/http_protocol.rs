//! HTTP-specific data structures.
//!
//! These mirror equivalent concepts from `prime_server`'s `http_protocol.hpp`.

use bitfield_struct::bitfield;
use tracing::warn;
use zerocopy::{LE, U16, U32};
use zerocopy_derive::{FromBytes, Immutable, IntoBytes, Unaligned};

/// HTTP request info.
///
/// In Valhalla service workers,
/// this is passed as the first message in any exchange sequence.
#[derive(FromBytes, IntoBytes, Immutable, Unaligned, Debug)]
#[repr(C)]
pub struct HttpRequestInfo {
    /// The request ID
    id: U32<LE>,
    /// The request timestamp.
    ///
    /// This SHOULD be an integer seconds UNIX timestamp
    /// (though it technically isn't specified very well).
    /// This is eventually going to run out of space (year-based bugs)
    /// or hit some bad behavior!
    timestamp: U32<LE>,
    inner_bitfield: HttpRequestInfoInnerBitfield,
    /// Required padding to make the struct 16-byte aligned.
    ///
    /// NB: This is implicit in the C++ struct definition.
    _spare: U16<LE>,
}

#[bitfield(u16,
    repr = U16<LE>,
    from = bit_twiddling_helpers::conv_u16le::from_inner,
    into = bit_twiddling_helpers::conv_u16le::into_inner
)]
#[derive(FromBytes, IntoBytes, Immutable, Unaligned)]
struct HttpRequestInfoInnerBitfield {
    /// Protocol-specific space for versioning info.
    ///
    /// Valhalla seems to use the following values:
    /// * 0 - HTTP/1.0
    /// * 1 - Technically anything else
    #[bits(3)]
    version: u8,
    /// Indicates whether the header is present or not
    #[bits(1)]
    connection_keep_alive: bool,
    /// Indicates whether the header is present or not
    #[bits(1)]
    connection_close: bool,
    /// What response code was set to when sent back to the client
    #[bits(10, from = bit_twiddling_helpers::conv_u16le::from_inner, into = bit_twiddling_helpers::conv_u16le::into_inner)]
    response_code: U16<LE>,
    /// Spare padding bits to be explicit about alignment
    #[bits(1)]
    _spare: u8,
}

impl HttpRequestInfo {
    /// The request ID (generated serially by Valhalla).
    pub fn id(&self) -> u32 {
        self.id.into()
    }

    /// The HTTP version string for this request (e.g. "HTTP/1.1").
    pub fn http_version_string(&self) -> &'static str {
        let version = self.inner_bitfield.version();
        match version {
            0 => "HTTP/1.0",
            1 => "HTTP/1.1",
            _ => {
                warn!(
                    "Unknown HTTP version from Valhalla service: {}; treating it as HTTP/1.1",
                    version
                );

                "HTTP/1.1"
            }
        }
    }

    pub fn connection_keep_alive(&self) -> bool {
        self.inner_bitfield.connection_keep_alive()
    }

    pub fn connection_close(&self) -> bool {
        self.inner_bitfield.connection_close()
    }

    pub fn set_response_code(&mut self, code: u16) {
        self.inner_bitfield.set_response_code(code.into())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use zerocopy::transmute;

    #[test]
    fn decode_http_info_message() {
        const MESSAGE: [u8; 12] = [
            0x00, 0x00, 0x00, 0x00, 0xf5, 0x76, 0xb1, 0x68, 0x01, 0x00, 0x00, 0x00,
        ];
        let parsed: HttpRequestInfo = transmute!(MESSAGE);

        if !cfg!(miri) {
            insta::assert_debug_snapshot!(parsed);
        }
    }
}
