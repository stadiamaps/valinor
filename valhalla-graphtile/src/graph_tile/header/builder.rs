use super::{
    FirstBitfield, GraphTileHeader, MiscCountsBitFieldOne, MiscCountsBitFieldTwo, SecondBitfield,
    TransitRecordBitfield, TransitionCountBitfield, TurnLaneCountBitfield, VALHALLA_EPOCH,
    VERSION_LEN,
};
use crate::graph_tile::predicted_speeds::COEFFICIENT_COUNT;
use crate::graph_tile::{
    AccessRestriction, Admin, DirectedEdge, DirectedEdgeExt, GraphTileBuildError, NodeInfo,
    NodeTransition, Sign, TransitDeparture, TransitRoute, TransitSchedule, TransitStop,
    TransitTransfer, TurnLane,
};
use crate::{BIN_COUNT, GraphId};
use chrono::{DateTime, Utc};
use geo::Coord;

/// A builder struct that helps us build correct graph tile headers.
///
/// This is not exposed in the public API by design at the moment,
/// because it is unstable and not as safe as it could be.
/// This can theoretically be made public if we figure out a way to make it a bit safer.
/// The main point of unsafety is that it requires the caller to self-report a lot of data
/// in header-friendly formats rather than user-friendly ones.
///
/// An alternate approach to the present would be moving most of the fields to the build method,
/// and requiring a dozen or more parameters there.
/// This feels a bit more unwieldy than a large struct init,
/// but perhaps it's the easiest way to make this safer long-term.
///
/// The main reason to make this public would be to enable use of the reserved fields.
/// However, we can probably figure out a way to enable that upstream in the [`crate::graph_tile::GraphTileBuilder`].
pub(crate) struct GraphTileHeaderBuilder {
    pub version: [u8; VERSION_LEN],
    pub graph_id: GraphId,
    /// The relative road density within this tile (0-15).
    ///
    /// NOTE: Valhalla doesn't typically set this when writing initial tiles.
    /// It sets them during a final validation step.
    pub density: u8,
    pub has_elevation: bool,
    pub has_ext_directed_edges: bool,
    pub sw_corner: Coord<f32>,
    /// The dataset ID (canonically, the OSM changeset ID).
    pub dataset_id: u64,
    pub node_count: usize,
    pub directed_edge_count: usize,
    pub predicted_speed_profile_count: usize,
    pub transition_count: usize,
    pub turn_lane_count: usize,
    pub transfer_count: usize,
    pub departure_count: usize,
    pub stop_count: usize,
    pub route_count: usize,
    pub schedule_count: usize,
    pub sign_count: usize,
    pub access_restriction_count: usize,
    pub admin_count: usize,
    pub create_date: DateTime<Utc>,
    // FIXME: This deserves a better interface
    pub bin_offsets: [u32; BIN_COUNT],
    /// Raw memory size in bytes
    pub complex_forward_restrictions_size: usize,
    /// Raw memory size in bytes
    pub complex_reverse_restrictions_size: usize,
    /// Raw memory size in bytes
    pub edge_info_size: usize,
    /// Raw memory size in bytes
    pub text_list_size: usize,
    /// Raw memory size in bytes
    pub lane_connectivity_size: usize,
}

impl GraphTileHeaderBuilder {
    pub(crate) fn build(self) -> Result<GraphTileHeader, GraphTileBuildError> {
        let bit_field_1 = FirstBitfield::default()
            // Guaranteed to be valid
            // (can theoretically be out of the accepted bounds, but this requires unsafe raw init,
            // and it won't corrupt the tile).
            .with_graph_id(self.graph_id.value().into())
            .with_density_checked(self.density.into())
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "density".to_string(),
                value: usize::from(self.density),
            })?
            // NOTE: name_quality, speed_quality, and exit_quality intentionally omitted
            // Guaranteed to be 0 or 1 by the Rust standard
            .with_has_elevation(self.has_elevation.into())
            .with_has_ext_directed_edge(self.has_ext_directed_edges.into());

        let bit_field_2 = SecondBitfield::default()
            .with_node_count_checked(u32::try_from(self.node_count)?.into())
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "node_count".to_string(),
                value: usize::from(self.node_count),
            })?
            .with_directed_edge_count_checked(u32::try_from(self.directed_edge_count)?.into())
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "directed_edge_count".to_string(),
                value: usize::from(self.directed_edge_count),
            })?
            .with_predicted_speeds_count_checked(
                u32::try_from(self.predicted_speed_profile_count)?.into(),
            )
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "predicted_speed_profile_count".to_string(),
                value: usize::from(self.predicted_speed_profile_count),
            })?;

        let transition_count_bitfield = TransitionCountBitfield::default()
            .with_transition_count_checked(u32::try_from(self.transition_count)?.into())
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "transition_count".to_string(),
                value: usize::from(self.transition_count),
            })?;

        let turn_lane_count_bitfield = TurnLaneCountBitfield::default()
            .with_turn_lane_count_checked(u32::try_from(self.turn_lane_count)?.into())
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "turn_lane_count".to_string(),
                value: usize::from(self.turn_lane_count),
            })?;

        let transit_record_bitfield = TransitRecordBitfield::default()
            .with_transfer_count_checked(u16::try_from(self.transfer_count)?.into())
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "transfer_count".to_string(),
                value: usize::from(self.transfer_count),
            })?
            .with_departure_count_checked(u32::try_from(self.departure_count)?.into())
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "departure_count".to_string(),
                value: usize::from(self.departure_count),
            })?
            .with_stop_count_checked(u16::try_from(self.stop_count)?.into())
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "stop_count".to_string(),
                value: usize::from(self.stop_count),
            })?;

        let misc_counts_bit_field_one = MiscCountsBitFieldOne::default()
            .with_route_count_checked(u16::try_from(self.route_count)?.into())
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "route_count".to_string(),
                value: usize::from(self.route_count),
            })?
            .with_schedule_count_checked(u16::try_from(self.schedule_count)?.into())
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "schedule_count".to_string(),
                value: usize::from(self.schedule_count),
            })?
            .with_sign_count_checked(u32::try_from(self.sign_count)?.into())
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "sign_count".to_string(),
                value: usize::from(self.sign_count),
            })?;

        let misc_counts_bit_field_two = MiscCountsBitFieldTwo::default()
            .with_access_restriction_count_checked(
                u32::try_from(self.access_restriction_count)?.into(),
            )
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "access_restriction_count".to_string(),
                value: usize::from(self.access_restriction_count),
            })?
            .with_admin_count_checked(u16::try_from(self.admin_count)?.into())
            .map_err(|_| GraphTileBuildError::BitfieldOverflow {
                field: "admin_count".to_string(),
                value: usize::from(self.admin_count),
            })?;

        // Calculate the starting offset of the variable sized data areas
        let dst_base_offset = size_of::<GraphTileHeader>()
            + (self.node_count * size_of::<NodeInfo>())
            + (self.transition_count * size_of::<NodeTransition>())
            + (self.directed_edge_count * size_of::<DirectedEdge>())
            + (if self.has_ext_directed_edges {
                self.directed_edge_count * size_of::<DirectedEdgeExt>()
            } else {
                0
            })
            + (self.access_restriction_count * size_of::<AccessRestriction>())
            + (self.departure_count * size_of::<TransitDeparture>())
            + (self.stop_count * size_of::<TransitStop>())
            + (self.route_count * size_of::<TransitRoute>())
            + (self.schedule_count * size_of::<TransitSchedule>())
            + (self.transfer_count * size_of::<TransitTransfer>())
            + (self.sign_count * size_of::<Sign>())
            + (self.turn_lane_count * size_of::<TurnLane>())
            + (self.admin_count * size_of::<Admin>())
            // Cross reference with edge_bins_size in header.rs...
            // The interface for this will improve at some point.
            + (usize::try_from(self.bin_offsets[BIN_COUNT - 1])? * size_of::<GraphId>());
        let complex_reverse_restrictions_offset =
            dst_base_offset + self.complex_forward_restrictions_size;
        let edge_info_offset =
            complex_reverse_restrictions_offset + self.complex_reverse_restrictions_size;
        let text_list_offset = edge_info_offset + self.edge_info_size;
        let lane_connectivity_offset = text_list_offset + self.text_list_size;
        let (predicted_speed_offset, tile_size) = if self.predicted_speed_profile_count > 0 {
            let pso = lane_connectivity_offset + self.lane_connectivity_size;
            let size = pso +
                // Offsets
                (self.directed_edge_count * size_of::<u32>())
                // Profile data
                + (self.predicted_speed_profile_count * size_of::<i16>() * COEFFICIENT_COUNT);
            (pso, size)
        } else {
            (0, lane_connectivity_offset + self.lane_connectivity_size)
        };

        Ok(GraphTileHeader {
            bit_field_1,
            base_lon_lat: [self.sw_corner.x.into(), self.sw_corner.y.into()],
            version: self.version,
            dataset_id: self.dataset_id.into(),
            bit_field_2,
            transition_count_bitfield,
            turn_lane_count_bitfield,
            transit_record_bitfield,
            misc_counts_bit_field_one,
            misc_counts_bit_field_two,
            _reserved_1: Default::default(),
            _reserved_2: Default::default(),
            complex_restriction_forward_offset: u32::try_from(dst_base_offset)?.into(),
            complex_restriction_reverse_offset: u32::try_from(complex_reverse_restrictions_offset)?
                .into(),
            edge_info_offset: u32::try_from(edge_info_offset)?.into(),
            text_list_offset: u32::try_from(text_list_offset)?.into(),
            create_date: ((self.create_date - VALHALLA_EPOCH).num_days().max(0) as u32).into(),
            bin_offsets: self.bin_offsets.map(|offset| offset.into()),
            lane_connectivity_offset: u32::try_from(lane_connectivity_offset)?.into(),
            predicted_speeds_offset: u32::try_from(predicted_speed_offset)?.into(),
            tile_size: u32::try_from(tile_size)?.into(),
            _empty_slots: [0.into(); _],
        })
    }
}
