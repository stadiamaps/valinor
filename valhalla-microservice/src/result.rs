use crate::http_protocol::HttpRequestInfo;
use http::header::{ACCESS_CONTROL_ALLOW_ORIGIN, CONNECTION, CONTENT_LENGTH, CONTENT_TYPE};
use http::{HeaderMap, StatusCode};
use itertools::intersperse;
use serde::Serialize;

/// The result of a worker computation.
pub enum WorkerResult {
    /// An HTTP response to be delivered to the user.
    ///
    /// No further processing will be done downstream.
    /// Use helper methods like [`WorkerResult::json`] wherever possible to avoid footguns,
    /// like forgetting to set headers.
    HttpResponse {
        status_code: StatusCode,
        /// A map of HTTP headers.
        ///
        /// If you are constructing this enum variant on your own (or implementing a new helper),
        /// be careful to set `Content-Type`, and any other headers specific to this type of response.
        ///
        /// `Content-Length`, `Access-Control-Allow-Origin`, and `Connection`
        /// may be set automatically during serialization; any values you set here will be clobbered.
        headers: HeaderMap,
        body: Vec<u8>,
    },
    // TODO: Figure out the other variants here...
    PlaceholderDownstreamTBD,
}

impl WorkerResult {
    /// Helper for constructing a JSON HTTP response.
    pub fn json<T: Serialize>(status_code: StatusCode, value: T) -> WorkerResult {
        let body = serde_json::to_vec(&value).expect("Programming error: either Serialize is incorrectly implemented, or the structure contains a map with non-string keys.");
        let mut headers = HeaderMap::new();
        headers.insert(
            CONTENT_TYPE,
            "application/json;charset=utf-8".parse().unwrap(),
        );

        WorkerResult::HttpResponse {
            status_code,
            headers,
            body,
        }
    }
}

pub(crate) fn serialize_http(
    request_info: HttpRequestInfo,
    status_code: StatusCode,
    headers: HeaderMap,
    body: Vec<u8>,
) -> Vec<u8> {
    let prelude: Vec<u8> =
        format!("{} {}\r\n", request_info.http_version_string(), status_code).into_bytes();
    let mut headers = headers;
    headers.insert(CONTENT_LENGTH, body.len().to_string().parse().unwrap());
    headers.insert(ACCESS_CONTROL_ALLOW_ORIGIN, "*".parse().unwrap());

    if request_info.connection_keep_alive() {
        headers.insert(CONNECTION, "Keep-Alive".parse().unwrap());
    } else if request_info.connection_close() {
        headers.insert(CONNECTION, "Close".parse().unwrap());
    }

    let headers: Vec<u8> = intersperse(
        headers
            .iter()
            .map(|(name, value)| [name.as_str().as_bytes(), b": ", value.as_bytes()].concat()),
        "\r\n".to_string().into_bytes(),
    )
    .flatten()
    .collect();

    [prelude, headers, "\r\n\r\n".to_string().into_bytes(), body].concat()
}

#[cfg(test)]
mod tests {
    use super::*;
    use zerocopy::transmute;

    #[test]
    fn test_serialize_http() {
        const BODY: &[u8] = br#"{"version":"3.5.1","tileset_last_modified":1756439278,"available_actions":["status","centroid","expansion","transit_available","trace_attributes","trace_route","isochrone","optimized_route","sources_to_targets","height","route","locate"]}"#;
        const REQ_INFO_BYTES: [u8; 12] = [
            0x00, 0x00, 0x00, 0x00, 0xf5, 0x76, 0xb1, 0x68, 0x01, 0x00, 0x00, 0x00,
        ];
        let req_info: HttpRequestInfo = transmute!(REQ_INFO_BYTES);

        let mut headers = HeaderMap::new();
        headers.insert(
            CONTENT_TYPE,
            "application/json;charset=utf-8".parse().unwrap(),
        );
        headers.insert(CONTENT_LENGTH, BODY.len().to_string().parse().unwrap());

        let result = serialize_http(req_info, StatusCode::OK, headers, BODY.to_vec());

        let reesult_utf8 = String::from_utf8(result).expect("Expected a UTF-8 result.");
        insta::assert_snapshot!(reesult_utf8);
    }
}
