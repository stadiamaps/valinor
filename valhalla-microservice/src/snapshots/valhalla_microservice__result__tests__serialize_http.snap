---
source: valhalla-microservice/src/result.rs
expression: reesult_utf8
---
HTTP/1.1 200 OK
content-type: application/json;charset=utf-8
content-length: 238
access-control-allow-origin: *

{"version":"3.5.1","tileset_last_modified":1756439278,"available_actions":["status","centroid","expansion","transit_available","trace_attributes","trace_route","isochrone","optimized_route","sources_to_targets","height","route","locate"]}
